README.md

```diff
@@ -1,2 +1,310 @@
+     * @accessToken External cloud access token
+     * @response {DiscoveryResponse} Discovery response object
+     */
+  })
+  .stateRefreshHandler((accessToken, response) => {
+    /**
+     * State refresh request. Respond with the current states of all devices. Called after
+     * device discovery runs.
+     * @accessToken External cloud access token
+     * @response {StateRefreshResponse} StateRefresh response object
+     */
+  })
+  .commandHandler((accessToken, response, devices) => {
+    /**
+     * Device command request. Control the devices and respond with new device states
+     * @accessToken External cloud access token
+     * @response {CommandResponse} CommandResponse response object
+     * @devices {array} List of ST device commands
+     */
+  })
+  .callbackAccessHandler((accessToken, callbackAuthentication, callbackUrls) => {
+    /**
+     * Create access and refresh tokens to allow SmartThings to be informed of device state
+     * changes as they happen. 
+     * @accessToken External cloud access token
+     * @callbackAuthentication ST access and refresh tokens for proactive state callbacks
+     * @callbackUrls Callback and refresh token URLs
+     */
+  })
+  .integrationDeletedHandler(accessToken => {
+    /**
+     * Called when the connector is removed from SmartThings. You may want clean up access
+     * tokens and other data when that happend.
+     * @accessToken External cloud access token
+     */
+  });
+
+```
+
+## Minimal loopback connector example
+This simple connector creates a one dimmer device named _Test Dimmer_. There's no physical
+device involved. The connector command handler simply returns the state value corresponding to 
+the issues command. The current state of the device is stored in memory, so if the server
+is restarted the states will revert to their initial value. This implementation does not 
+implement proactive state callbacks.
+
+#### connector.js
+```javascript
+const {SchemaConnector} = require('st-schema')
+const deviceStates = { switch: 'off', level: 100}
+const connector = new SchemaConnector()
+  .discoveryHandler((accessToken, response) => {
+    response.addDevice('external-device-1', 'Test Dimmer', 'c2c-dimmer')
+  })
+  .stateRefreshHandler((accessToken, response) => {
+    response.addDevice('external-device-1', [
+      {
+        component: 'main',
+        capability: 'st.switch',
+        attribute: 'switch',
+        value: deviceStates.switch
+      },
+      {
+        component: 'main',
+        capability: 'st.switchLevel',
+        attribute: 'level',
+        value: deviceStates.level
+      }
+    ])
+  })
+  .commandHandler((accessToken, response, devices) => {
+    for (const device of devices) {
+      response.addDevice(device.externalDeviceId, device.commands.map( cmd => {
+        if (cmd.command === 'setLevel') {
+          deviceStates.level = cmd.arguments[0];
+          return {
+            component: cmd.component,
```

Comment:
since only attribute and value is different would it make sense to create a default response outside the `if`, assign attribute/value inside. Granted the return value has to be cleared in case the command doesn't match.

Reply:
Yeah, that's better approach in the face of possibly more commands, but a little less clear in a sense. You have to look a multiple places in the code and construct what the complete response looks like in your head, rather than seeing it written out. I'll play around with that. 

Regarding commands that don't match, what is the recommended way of handling that? The platform independently checks that the command is defined for the capability, and rejects it if it isn't, but what about the case in which the connector doesn't handle a command that is part of the capability. I hesitate to say the whole thing should fail, since there can be multiple commands in one request. The closes thing I see in the docs is the `CAPABILITY-NOT-SUPPORTED` response:
```
{
      "externalDeviceId": "pdevice-1",
      "deviceError": [
        {
          "errorEnum": "CAPABILITY-NOT-SUPPORTED",
          "detail": "more detail from the partner"
        }
      ]
}
```

That's not necessarily the problem exactly, but close. Is that the recommended behavior?

Reference:
https://smartthings.developer.samsung.com/docs/guides/smartthings-schema/smartthings-schema-reference.html

Reply:
:thumbsup: made sense to do that along with the error handling. I'm still wondering about the situation where there are multiple commands associated with a request, and some are OK and other aren't. Will returning an error cause valid state updates for the same device to be ignored?